File: leetcode_ly/c++/BackTracking/GenerateParentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        string s = "";
        vector<string> result;
        if(n == 0) return result;
        getResult(s,result,n,n);
        return result;
    }
    
    void getResult(string s,vector<string>& result,int left,int right){
        if(left == 0 && right == 0){
            result.push_back(s);
            return;
        }
        
        if(left > 0){
            getResult(s + '(',result,left - 1,right);
            s.substr(0,s.size() - 1);
        }
        
        if(left < right){
            getResult(s + ')',result,left,right - 1);
            s.substr(0,s.size() - 1);
        }
    }
};

File: leetcode_ly/java/String/GenerateParentheses.java
public class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<String>();
        if(n == 0) return res;
        generate(res,"",n,n);
        return res;
    }
    
    public void generate(List<String> res,String ans,int left,int right){
        if(left > right) return;
        if(left == 0 && right == 0){
            res.add(ans);
            return;
        }
        
        if(left > 0){
            ans = ans + '(';
            generate(res,ans, left - 1,right);
            ans = ans.substring(0,ans.length() - 1);
        }
        if(right > 0){
            ans = ans + ')';
            generate(res,ans, left,right - 1);
            ans = ans.substring(0,ans.length() - 1);
        }
    }
}